### PR TITLE
Initialize event channels early

### LINKED ISSUE #137: ios subscribeToCharacteristic crashes the app
**Describe the bug**
Subsrcibing to specific characteristic crashes ios app immediately. Android one works well with the same code.

**To Reproduce**
1. use sample app 
2. change body of the function _Future<void> discoverServices(String deviceId) async_ 
to something like 
`final characteristic = QualifiedCharacteristic(serviceId: Uuid.parse('4c0719b0-c810-492b-87b1-63c6c3fd584f'), characteristicId: Uuid.parse('4c0719b3-c810-492b-87b1-63c6c3fd584f'), deviceId: deviceId);
    
      _ble.subscribeToCharacteristic(characteristic).listen((data) {
        print(data.toString());
      }, onError: (dynamic error) {
        print(error);
      });`
3. launch the app
4. tap search button
5. select device
6. tap connect and wait until status changes
7. tap Discover services
8. app immediately crashes - tries to subsrcibe to characteristic

Note: original code for discovering services works properly on both ios and android, so service and characteristics are discovered
.

**Expected behavior**
Should receive data printed on console - Android works properly

**Smartphone / tablet**
 - Device:iPhone XS
 - OS: iOS 14
 - Package version: 2.6.1+1

Attached screenshot from xcode
Screenshot 2020-11-04 at 23 42 18

### PR DESCRIPTION
**Pull request by werediver**
Fixes #137

Also corrects an error message mentioned in #138. @HelloBytes Thank you for uncovering it!

### CODE REVIEW

File: ios/Classes/ReactiveBle/Central.swift [hunk 1]
```diff
@@ -337,7 +337,7 @@ final class Central {
                 return "A peripheral \(peripheralID.uuidString) is unknown (make sure it has been discovered)"
             case .peripheralIsNotConnected(let peripheralID):
                 return "The peripheral \(peripheralID.uuidString) is not connected"
-            case .serviceNotFound(let peripheralID, let serviceID):
+            case .serviceNotFound(let serviceID, let peripheralID):
```
Comment (werediver) [inline ios/Classes/ReactiveBle/Central.swift:340]: This relates to #138.

File: lib/src/connected_device_operation.dart [hunk 2]
```diff
@@ -43,7 +43,7 @@ class ConnectedDeviceOperation {
 
   Stream<List<int>> subscribeToCharacteristic(
     QualifiedCharacteristic characteristic,
-    Future<ConnectionStateUpdate> shouldTerminate,
+    Future<void> isDisconnected,
```
Comment (werediver) [inline lib/src/connected_device_operation.dart:46]: A cosmetic change. This naming makes it much clearer, _why_ to stop the subscription when this future complete.

File: lib/src/plugin_controller.dart [hunk 3]
```diff
@@ -19,10 +19,10 @@ class PluginController {
...
     @required ProtobufConverter protobufConverter,
     @required MethodChannel bleMethodChannel,
-    @required EventChannel connectedDeviceChannel,
-    @required EventChannel charUpdateChannel,
-    @required EventChannel bleDeviceScanChannel,
-    @required EventChannel bleStatusChannel,
+    @required Stream<List<int>> connectedDeviceChannel,
+    @required Stream<List<int>> charUpdateChannel,
+    @required Stream<List<int>> bleDeviceScanChannel,
+    @required Stream<List<int>> bleStatusChannel,
```
Comment (werediver) [inline lib/src/plugin_controller.dart:25]: This is a bit lazy. I'd prefer to inject _fully decoded_ streams (as we do in Hue Bluetooth), but this half-way transition is faster to implement. We can improve on it later, I guess.

File: lib/src/plugin_controller.dart [hunk 4]
```diff
@@ -313,10 +299,14 @@ class PluginControllerFactory {
...
-      bleDeviceScanChannel: scanEventChannel,
-      bleStatusChannel: bleStatusChannel,
+      connectedDeviceChannel:
+          connectedDeviceChannel.receiveBroadcastStream().cast<List<int>>(),
+      charUpdateChannel:
+          charEventChannel.receiveBroadcastStream().cast<List<int>>(),
+      bleDeviceScanChannel:
+          scanEventChannel.receiveBroadcastStream().cast<List<int>>(),
+      bleStatusChannel:
+          bleStatusChannel.receiveBroadcastStream().cast<List<int>>(),
```
Comment (werediver) [inline lib/src/plugin_controller.dart:309]: Now event channel initialization is not lazy, which should fix #137, I reckon.

File: lib/src/reactive_ble.dart [hunk 5]
```diff
@@ -317,19 +317,20 @@ class FlutterReactiveBle {
...
+    QualifiedCharacteristic characteristic,
+  ) {
+    final isDisconnected = connectedDeviceStream
         .where((update) =>
             update.deviceId == characteristic.deviceId &&
             (update.connectionState == DeviceConnectionState.disconnecting ||
                 update.connectionState == DeviceConnectionState.disconnected))
-        .firstWhere((_) => true,
-            orElse: () => throw NoBleCharacteristicDataReceived());
+        .cast<void>()
```
Comment (werediver) [inline lib/src/reactive_ble.dart:327]: We don't need any values where it goes, so I'm erasing the type here.

File: test/plugin_controller_test.dart [hunk 6]
```diff
@@ -20,28 +20,28 @@ void main() {
...
       _methodChannel = _MethodChannelMock();
       _protobufConverter = _ProtobufConverterMock();
-      _connectedDeviceChannel = _EventChannelMock();
-      _argsChannel = _EventChannelMock();
-      _scanChannel = _EventChannelMock();
-      _statusChannel = _EventChannelMock();
+      _connectedDeviceStreamController = StreamController();
+      _argsStreamController = StreamController();
+      _scanStreamController = StreamController();
+      _statusStreamController = StreamController();
```
Comment (werediver) [inline test/plugin_controller_test.dart:35]: To my taste it's nice that we don't have to mess with (even mock) event channels in the tests.

File: lib/src/connected_device_operation.dart (hunk 2 above)
Comment (remonh87) [reply lib/src/connected_device_operation.dart:46]: much better indeed

**Review by remonh87 — APPROVED** [review]
Much better indeed like the improvement that we do not need to mock event channels in our tests

**Pull request merged**